docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,350 @@
+*/
+int sgx_mm_uncommit(void *addr, size_t length);
+
+/**
+ * Deallocate the address range
+ * The pages in the allocation are freed and the address range is released for future allocation
+ * @param[in] addr page aligned start address of the region to be freed and released
+ * @param[in] length size in bytes of multiples of page size.
+ * @return 0 on success, EINVAL if address range is not allocated previously, -1 on other failures 
+ */
+int sgx_mm_dealloc(void* addr, size_t length);
+
+```
+
+### sgx_mm_modify_type, sgx_mm_modify_permissions and sgx_mm_modify_ex
+
+```
+/**
+ * Change permissions and/or page type of a previously allocated region
+ * @param[in] addr start address of the region, must be page aligned
+ * @param[in] length size in bytes of page multiples.
+ * @param[in] prot permissions bitwise OR of following with the same values as those defined for
+ * Linux mmap/mprotect syscalls:
+ *        - PROT_READ: Pages may be read.
+ *        - PROT_WRITE: Pages may be written. 
+ *        - PROT_EXECUTE: Pages may be executed.
+ *        
+ * @param[in] type page type, one of following:
+ *       - PT_TCS: TCS page
+ *       - PT_SS_FIRST: the first page in shadow stack
+ *       - PT_SS_REST: the rest page in shadow stack
+ *       - -1: no page type change, keep the original type
+ * 
+ * @returns 0 on success, otherwise EINVAL if the memory region was not previously allocated, or other
+ * invalid parameters,  EACCES if original page type can not be changed to target type, EPERM if the request permissions
+ * are not allowed (e.g., not allowed by target page type or SELinux policy), or target page type 
+ * is no allowed by this call, e.g., PT_TRIM, -1 otherwise. 
+ */
+int sgx_mm_modify_ex(void *addr, size_t length, int prot, int type);
+
+/**
+ * Change permissions of an allocated region
+ * Equivalent to sgx_mm_modify_ex(addr, length, prot, -1) 
+ */
+int sgx_mm_modify_permissions(void *addr, size_t length, int prot);
+/**
+ * Change permissions of an allocated region
+ * Equivalent to sgx_mm_modify_ex(addr, length, -1, type) 
+ */
+int sgx_mm_modify_type(void *addr, size_t length, int type);
+
+```
+**Remarks:**
+- The memory manager will track current permissions for each region, and can determine whether new permission requires OCALL for EMODPR, e.g RW<->RX, RW->R
+- This API should not be used to change EPC page type to PT_TRIM. Trimming pages are done by sgx_mm_uncommit and sgx_mm_dealloc only.
+
+
+### sgx_mm_commit
+```
+/**
+* Commit a partial or full range of memory allocated previously with EMA_COMMIT_ON_DEMAND.
+* The API will return 0 if all pages in the requested range are successfully committed. 
+* Calling this API on pages already committed has not effect. 
+* @param[in] addr, page aligned starting address
+* @param[in] length, length of the region in bytes of multiples of page size.
+* @return 0 on success. EINVAL if any requested page is not in any previously allocated regions, -1 otherwise 
```

Comment:
Why return -1 for other failures instead of a more specific error code from https://linux.die.net/man/3/errno or from `oe_result_t`?

Reply:
updated to return EFAULT for all other errors